### PR TITLE
fix all_gather_into_tensor test and logic

### PR DESCRIPTION
**Pull request by bfolie**
Most cases now work, except for stacking + compilation, which requires a change to upstream pytorch (https://github.com/pytorch/pytorch/issues/155632)

### CODE REVIEW

File: torch_xla/distributed/xla_backend.py [hunk 1]
```diff
@@ -82,15 +82,33 @@ def allreduce(self, tensors, all_reduce_options):
     return _ret_work(tensors)
 
   # method for dist.all_gather_into_tensor under eager mode.
-  def _allgather_base(self, output_tensor, input_tensor, opts):
-    return self.allgather(output_tensor, input_tensor, opts)
+  def _allgather_base(self, output_tensor: torch.Tensor,
+                      input_tensor: torch.Tensor, opts: AllgatherOptions):
+    is_scalar = (input_tensor.dim() == 0)
```
Comment (bhavya01) [inline torch_xla/distributed/xla_backend.py:88]: Do you think it'll help if we point to pytorch documentation for all_gather_into_tensor here?

**Review by bhavya01 — APPROVED** [review]

File: test/pjrt/test_collective_ops_tpu.py [hunk 2]
```diff
@@ -167,19 +167,25 @@ def callable(input):
...
     dist.init_process_group("xla", init_method='xla://')
     device = torch_xla.device()
     input = torch.tensor([xr.global_ordinal()],
                          dtype=torch.float,
                          device=device)
-    output_tensor = torch.empty((1, xr.world_size()), device=device)
+    if mode == "stack":
+      output_tensor = torch.empty((xr.world_size(), 1), device=device)
+    elif mode == "concat":
+      output_tensor = torch.empty((xr.world_size(),), device=device)
```
Comment (pgmoka) [inline test/pjrt/test_collective_ops_tpu.py:185]: Above we have a check for `if mode not in allowed_modes:`. I think this should be moved there, and have the error be thrown in an `else` statement. You could also have only `(xr.world_size(), 1)` or `(xr.world_size(), 0)` be selected. Example:
```
  def _all_gather_into_tensor(use_dynamo: bool, mode: str):
    met.clear_all()
    if mode == "stack":
      world_size = (xr.world_size(), 1)
    elif mode == "concat":
      world_size = (xr.world_size(), 0)
    else:
      raise ValueError(f"mode must be one of {["stack", "concat"]}")
      
  ...
  
  output_tensor = torch.empty(world_size, device=device)
```

File: torch_xla/distributed/xla_backend.py [hunk 3]
```diff
@@ -82,15 +82,33 @@ def allreduce(self, tensors, all_reduce_options):
...
+    else:
+      stacked_result = torch.stack(
+          torch.split(result, input_tensor.shape[0], dim=0), dim=0)
+      if stacked_result.shape == output_tensor.shape:
+        output_tensor.copy_(stacked_result, non_blocking=True)
+      else:
+        msg = f"Input shape {input_tensor.shape} and output shape {output_tensor.shape} are not compatible for all_gather_into_tensor. Input must be stacked or concatenated to create output."
+        raise ValueError(msg)
+
+    return _ret_work([output_tensor])
```
Comment (pgmoka) [inline torch_xla/distributed/xla_backend.py]: Readability suggestion: using early returns can decrease if branching. Here you could consider something like:

```
 def _allgather_base(self, output_tensor: torch.Tensor,
                      input_tensor: torch.Tensor, opts: AllgatherOptions):
    is_scalar = (input_tensor.dim() == 0)
    if is_scalar:
      input_tensor = torch.reshape(input_tensor, (1,))

    result = xm.all_gather(
        input_tensor, dim=0, groups=self._mesh, pin_layout=False)

    if result.shape == output_tensor.shape:
      output_tensor.copy_(result, non_blocking=True)
      return _ret_work([output_tensor])

    stacked_result = torch.stack(
        torch.split(result, input_tensor.shape[0], dim=0), dim=0)
    if stacked_result.shape != output_tensor.shape:
      msg = f"Input shape {input_tensor.shape} and output shape {output_tensor.shape} are not compatible for all_gather_into_tensor. Input must be stacked or concatenated to create output."
      raise ValueError(msg)

    output_tensor.copy_(stacked_result, non_blocking=True)
    return _ret_work([output_tensor])
```

File: torch_xla/distributed/xla_backend.py [hunk 4]
```diff
@@ -82,15 +82,33 @@ def allreduce(self, tensors, all_reduce_options):
...
-  def _allgather_base(self, output_tensor, input_tensor, opts):
-    return self.allgather(output_tensor, input_tensor, opts)
+  def _allgather_base(self, output_tensor: torch.Tensor,
+                      input_tensor: torch.Tensor, opts: AllgatherOptions):
+    is_scalar = (input_tensor.dim() == 0)
+    if is_scalar:
+      input_tensor = torch.reshape(input_tensor, (1,))
+    result = xm.all_gather(
+        input_tensor, dim=0, groups=self._mesh, pin_layout=False)
+    if result.shape == output_tensor.shape:
```
Comment (pgmoka) [inline torch_xla/distributed/xla_backend.py]: NIT: Add space between if statements

File: test/pjrt/test_collective_ops_tpu.py (hunk 2 above)
Comment (bfolie) [reply test/pjrt/test_collective_ops_tpu.py:185]: I combined the output_tensor creation and value checking, but find separating the definition of the shape and the creation of output_tensor to be needlessly clever. I'd rather have the connection between the mode and resulting output be in the same place

Comment (pgmoka) [reply test/pjrt/test_collective_ops_tpu.py:185]: I think that is fair. Perhaps than something like
```
    if mode == "stack":
      output_tensor = torch.empty((xr.world_size(), 1), device=device)
    else # mode == "concat":
      output_tensor = torch.empty((xr.world_size(),), device=device)
```
or
```
    output_tensor = torch.empty((xr.world_size(),), device=device)
    if mode == "stack":
      output_tensor = torch.empty((xr.world_size(), 1), device=device)
```

Might be helpful to avoid unintended edge cases.

File: test/pjrt/test_collective_ops_tpu.py [hunk 5]
```diff
@@ -278,13 +284,17 @@ def test_all_reduce(self, use_dynamo):
     for index, val in results.items():
       torch.testing.assert_close(val, expected)
 
-  @parameterized.named_parameters(('dynamo', True), ('nondynamo', False))
-  def test_all_gather_into_tensor(self, use_dynamo):
+  @parameterized.product(dynamo=[True, False], mode=["stack", "concat"])
```
Comment (pgmoka) [inline test/pjrt/test_collective_ops_tpu.py:287]: Minor suggestion: It might be nice to leverage testing as documentation and have somewhere explain the difference between the two modes.

File: test/pjrt/test_collective_ops_tpu.py [hunk 6]
```diff
@@ -278,13 +284,17 @@ def test_all_reduce(self, use_dynamo):
...
+  def test_all_gather_into_tensor(self, dynamo, mode):
```
Comment (pgmoka) [inline test/pjrt/test_collective_ops_tpu.py:288]: NIT: It would help to make this test more readable if there was a space between logical if statements:
```
    if dynamo and mode == "stack":
      self.skipTest("https://github.com/pytorch/pytorch/issues/155632")
      
    results = pjrt.run_multiprocess(
        self._all_gather_into_tensor, use_dynamo=dynamo, mode=mode)
    expected = torch.arange(
        tpu.num_expected_global_devices(), dtype=torch.float)
        
    if mode == "stack":
      expected = expected.unsqueeze(1)
      
    for _, val in results.items():
      torch.testing.assert_close(val, expected)
```

**Review by pgmoka — APPROVED** [review]
LGTM minus some very small nits and suggestions.
